modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/AnnotatedFieldsWithDefaultConstructor.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.marshaller;
+
+import static java.lang.invoke.MethodType.methodType;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Constructor;
+import org.apache.ignite.lang.MarshallerException;
+
+class AnnotatedFieldsWithDefaultConstructor implements Creator.Annotated {
+    private final MethodHandle mhNoArgs;
+
+    AnnotatedFieldsWithDefaultConstructor(Constructor<?> defaultCtor) {
+        assert defaultCtor.getParameterCount() == 0;
+        this.mhNoArgs = unreflect(defaultCtor);
+    }
+
+    @Override
+    public Object createFrom(FieldAccessor[] accessors, MarshallerReader reader) {
+        try {
+            Object instance = mhNoArgs.invokeExact();
+
+            for (int fldIdx = 0; fldIdx < accessors.length; fldIdx++) {
+                accessors[fldIdx].read(reader, instance);
+            }
```

Comment:
Can you please comment on why you're expecting that the order in `accessors` will match the order from `reader`? Maybe there are some invariants that you forgot to document?